### DISCUSSION TITLE
Using async engine with ORM unable to execute basic queries

### DISCUSSION THREAD

**Discussion by fleetofthought** [sqlalchemy/sqlalchemy#7331] [Usage Questions]
I have switched my SQLAlchemy database code to use an async engine and am having trouble establishing basic functionality.

Right now, all I am trying to achieve is to simply select all tables in the database.

I have a class that initiates the database like this:
```
class PostgresDb:

    def __init__(self):
        self._session = None
        self._engine = None

    def __getattr__(self, name):
        return getattr(self._session, name)

    def init(self):
        self._engine = create_async_engine(
            ENGINE,
            echo=True,
            future=True)

        self._session = sessionmaker(
            self._engine, expire_on_commit=False, class_=AsyncSession
        )()

   async def create_all(self):
        async with self._engine.begin() as conn:
            await conn.run_sync(Base.metadata.create_all)

   # Other methods...
```

Example of how create_all gets called:
```
async def init_db_tables():
    self.init()
    await self.create_all()
asyncio.run(init_db_tables())
```

When I want to achieve basic functionality, like getting all the tables, I can do something like:
```
def get_tables(self):
    with create_engine(SYNCHRONOUS_ENGINE).connect() as conn:
        meta = MetaData(conn, schema=SCHEMA)
        meta.reflect(views=True)
        table_list = meta.tables
        return table_list
```
This is not ideal as I need to actually pass in a synchronous engine connection instead of the actual async engine I am using in the Class. It is also very verbose and shouldn't need to be initiated like this for every query.

I have tried doing something like this to select the table 'appuser' from the database:
```
async def get_tables(self):
    self.init()
    async with self._session() as session:
        q = select('appuser')
        result = await session.execute(q)
        curr = result.scalars()
        for i in curr:
            print(i)
```
Which I've tried calling like this
```
db = PostgresDb()
asyncio.run(db.get_tables())
asyncio.get_event_loop().run_until_complete(db.get_tables())
```
These both give error:
```
async with self._session() as session:
TypeError: 'AsyncSession' object is not callable
```

Calling it with `db.get_tables()` errors ```RuntimeWarning: coroutine 'PostgresDb.get_tables' was never awaited
  db.get_tables() RuntimeWarning: Enable tracemalloc to get the object allocation traceback```

Trying to use `inspector` with `run_sync` like this:
```
    async def get_tables(self):
        await self.init()
        async with self._engine.begin() as conn:
            inspector = conn.run_sync(inspect(conn))
            table_names = await conn.run_sync(inspector.get_table_names())
            print(table_names)
```
Returns error
```
sqlalchemy.exc.NoInspectionAvailable: Inspection on an AsyncConnection is currently not supported. Please use ``run_sync`` to pass a callable where it's possible to call ``inspect`` on the passed connection. 
```

I have read the documentation at https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncConnection.run_sync but I am still unclear about how to work cleanly with async engines.

Thanks for any and all insight you're able to offer on how to execute a simple query `get all tables` in SQLAlchmey using the async engine!

**Comment by zzzeek** [comment]
so far the immediate problem you have is that you are instantating the session:

    self._session = sessionmaker(
            self._engine, expire_on_commit=False, class_=AsyncSession
        )()   # <--- note extra ()

so that when you call  "async with self.session()" you already have an AsyncSession there and that will lead to all kinds of issues.

for the run_sync part, see the example at https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#using-the-inspector-to-inspect-schema-objects , note:

    def use_inspector(conn):
        # do things with sync conn
        inspector = inspect(conn)
        # use the inspector

and **not** this, which won't work

    inspector = conn.run_sync(inspect(conn))
    table_names = await conn.run_sync(inspector.get_table_names())

above wont work because "conn" is an AsyncConnection, you can't pass that to inspect().   run_sync accepts a *callable* that will be passed the sync connection you are looking for.
